Add various shutil methods

martinhoefling opened the pull request:
Added various methods like copy2, copytree etc. Also added a kwarg that allows to perform copying server side. Many methods are copied and adapted from their python2 shutil counterpart to mimic their behavior. Further, a set of integration tests that at least cover the happy paths was added. @jborean93 let me know what you think.

jborean93 commented:
Awesome work thanks for doing this. I'm still working through some final stuff on the easy-api branch and once that is in I'll start the review on this so we can include it in the same release.

martinhoefling commented:
@jborean93 all right, I will try to fix the build in the meantime.

martinhoefling commented:
Some parts still require tests and I am not sure how to handle Py2/3 compatibility for some of the error formattings. I'll add some more tests for the missing branches.

martinhoefling commented:
Ah, this is actually a pytest bug when `pytest.raises` tries to match a unicode string. https://github.com/pytest-dev/pytest/issues/5478. I'll do the strict checking for the message only on python3.

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+    SMB2SrvRequestResumeKey
+from smbprotocol.open import FilePipePrinterAccessMask, CreateOptions, ShareAccess
+
+CHUNK_SIZE = 1 * 1024 * 1024  # maximum chunksize allowed by smb
```
Where did you get this value from? SMB has a max packet size of around 65KB but newer dialects can go even larger than this.

martinhoefling replied:
This has nothing to do with the smb packet size (3.3.3 in the attached document). If I read this correctly, the widely used default is 1M per chunk max and max 16MB in chunks = 16 chunks of 1M. The 16 is hidden, I should extract it as variable as well.
[[MS-SMB].pdf](https://github.com/jborean93/smbprotocol/files/3833275/MS-SMB.pdf)

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+
+
+def _samefile(src, dst):
+    return normcase(abspath(src)) == normcase(abspath(dst))
```
If you want a proper samefile then `smbclient.path` has a function that actually compares the file id on the host but it is only really designed for remote -> remote comparisons.

martinhoefling replied:
I have extended samefile a bit to deal with certain scenarios.

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+def copyfile(src, dst, **kwargs):
+    """Copy data from src to dst"""
+    if _samefile(src, dst):
+        raise Error("`%s` and `%s` are the same file" % (src, dst))
```
Better as `ValueError`

martinhoefling replied:
done

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+            pass
+        else:
+            # XXX What about other special files? (sockets, devices...)
+            if py_stat.S_ISFIFO(st.st_mode):
```
We don't set this value at all, I've tried to find a way to determine the type of remote object, like a named pipe, outside of files/dirs/reparse points but have never really found a proper way for that.

martinhoefling replied:
I removed the entire check.

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+                raise SpecialFileError("`%s` is a named pipe" % fn)
+
+    with open_file(src, 'rb', **kwargs) as fsrc:
+        with open_file(dst, 'wb', **kwargs) as fdst:
```
`with` supports multiple entries so this could just be `with open_file(src, 'rb', **kwargs) as fsrc, open_file(dst, 'wb', **kwargs) as fdst:`. https://docs.python.org/3/reference/compound_stmts.html#the-with-statement

martinhoefling replied:
non server side copy code removed for now

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+    st = stat(src, **kwargs)
+    utime(dst, ns=(st.st_atime_ns, st.st_mtime_ns), **kwargs)
+
+    # TODO: in principle, we can copy the readonly flag via chmod
```
This is partly why I don't think we should ship with the copy2 functionality for now. Better to focus on getting a basic file copy that supports all the permutations of local/remote for now.

martinhoefling replied:
removed

jborean93 commented:
Also, if you rebase the commit against easy-api I've removed Python 3.4 from the matrix as it is EOL and I've removed support for it.

martinhoefling commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2019, Jordan Borean (@jborean93) <jborean93@gmail.com>
```
removed

martinhoefling commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+        def onerror(*args):
+            raise
+    try:
+        if islink(path, **kwargs):
```
removed the link check

martinhoefling commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+
+    names = []
+    try:
+        names = listdir(path, **kwargs)
```
I've written it according to your example in conftest with catching errors as rmtree does.

martinhoefling commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+        onerror(rmdir, path, sys.exc_info())
+
+
+def copy2(src, dst, server_side_copy=False, **kwargs):
```
So what we're interested in is mainly server side copy of files and entire directory trees. Local <-> remote was not considered at all, your're right. We could add this, by checking the UNC paths and if it's a UNC path at all. If I understand your suggestion correctly, you still wouldn't call that "copy2", because of more things listed above, that need to be taken into consideration?
Maybe adding that points to the copy2 documentation is an alternative? The python documentation does it that way for example (https://docs.python.org/3/library/shutil.html). I also have no problem in renaming the functions to reserve the name for a more complete one in the future.

jborean93 replied:
I think we should have `copy` at a minimum and potentially look at `copyfileobj` snf `copyfile`. The `copy2`, `copymode`, and `copystat` should be reserved for the future when we have a clearer picture and more time to really think about how to implement it here. Due to the complexity I think it should be done in a separate PR and not as part of this basic implementation.

As for the local <-> remote question, I'm happy to just have remote <-> remote in this PR for now but remove the `server_side_copy` option. We can still raise a ValueError if the path isn't remote and just remove that in a future iteration. I don't think it warrants a `copy2` consideration, it still fits under the `copy` spec where it's just copying the file's metadata, smbclient just allows it to be done over SMB :).

martinhoefling replied:
OK got it. Just the point with server_side_copy is not clear to me. Let's talk just about remote copy for a moment: Here, we can only do server_side_copy, if the server supports it (e.g. recent SAMBA versions) and we're on the same share/server. If not, we have to fall back to reading and writing chunks from python. Not sure how to find out if the server supports it though. That's why I've introduced the flag to indicate, that the server supports it (default is False).

The python chunk copy (which requires up and download of the files) can be easily reinforced to also work with local files.

jborean93 replied:
Hmm ok I wasn't aware of that, looking at the Samba docs I can see that it introduced support for `FSCTL_SRV_COPYCHUNK` in 4.1.0 https://wiki.samba.org/index.php/Server-Side_Copy which has been EOL since 2016 https://wiki.samba.org/index.php/Samba_Release_Planning.

It's hard to tell when it was supported on the server side for Windows but considering

1. https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/eaa129d1-fca4-46b1-8ccd-b91f3e1d2d80#Appendix_A_Target_132 doesn't mention a minimum dialect version
2. https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb/c82ff91d-28d8-4274-a2ab-c2f5ade26b98 - It is supported in SMB1

I would say it has been supported on Windows for a while now.

Because of this I feel like we can rely on the fact that the server supports the server side copy operations and we don't need to fall back on our own manual read/write of chunks. If someone was using the ancient version of Samba I don't really want to make things easier for them (reason why I didn't add SMB1 support). As for whether it is still remote <-> remote but different server/share then we can just check the path to make sure they are the same before using the IOCTL commands or falling back to our own manual chunk read/writing.

Lastly, this isn't necessary but it sounds like the pure server/share copy through IOCTL should be a function in `smbclient._os` and `smbclient.shutil` just calls that if it's able to. The `._os` function is designed to expose the pure SMB side in an easy to use interface which this sounds like a good candidate. To better explain what I mean

```
# smbclient._os.py
def copy(src, dst, **kwargs):
    check that src and dst are on the same share
    Call IOCTL FSCTL_SRV_COPYCHUNK to copy the files


# smbclient.shutil.py
def copy(src, dst, **kwargs):
    if src and dst are on the same share
        call _os.py's copy(src, dst, **kwargs)
    else
        manually read and write from src/dst ourselves
```

martinhoefling replied:
I have renamed it and removed all the functionality to copy between local and shares or between shares for now.

martinhoefling commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,306 @@
+    else:
+        raise ValueError("Target %s already exists" % norm_dst)
+
+    with SMBRawIO(norm_src, mode='r', desired_access=FilePipePrinterAccessMask.GENERIC_READ,
```
After rebase, the test are failing, I'll have to find out why first.

jborean93 replied:
Some of the failures were my doing sorry, I renamed some kwargs to avoid the builtin names. One I know off the top of my head is that `ioctl_request's` `buffer` kwarg is now `input_buffer`.

martinhoefling replied:
Right, that was the issue!

martinhoefling replied:
When changing to the suggested `open_file` I get
```
>       sid = self.raw.fd.tree_connect.session.session_id
E       AttributeError: '_io.TextIOWrapper' object has no attribute 'fd'
```

jborean93 replied:
It sounds like you need to open it in binary mode. Typically you will want to set `mode='<x>b` (the b is important) to make sure you are dealing with bytes. Internal functions should avoid using the text side unless they really need it.

martinhoefling commented:
I have cleaned up the code a bit, rmtree, copy and copytree_copy are in shutil, the server side implementation has been moved to _os. I still wasn't able to make the SMBRawIO replacement working. I will check out next week.

jborean93 commented on smbclient/_os.py:
```diff
@@ -55,7 +55,7 @@
     CtlCode,
     IOCTLFlags,
-)
+    SMB2SrvCopyChunk, SMB2SrvCopyChunkResponse, SMB2SrvCopyChunkCopy, SMB2SrvRequestResumeKey)
```
Best to keep the same style as above and put an import on a newline. This is mostly to make it easier to alphabetise and add/remove imports without worrying about the line length

martinhoefling replied:
done

jborean93 commented on smbclient/_os.py:
```diff
@@ -75,10 +75,11 @@
 XATTR_CREATE = getattr(os, 'XATTR_CREATE', 1)
 XATTR_REPLACE = getattr(os, 'XATTR_REPLACE', 2)
 
+MAX_COPY_CHUNK_SIZE = 1 * 1024 * 1024  # maximum chunksize 1M from 3.3.3 in MS-SMB documentation
```
I'm still not sure where this size comes from? MS-SMB is the SMBv1 implementation which is a different protocol from SMBv2 (MS-SMB2). I also can't find mention of this in the docs on MS-SMB2 talking about receiving this request at https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/cd0162e4-7650-4293-8a2a-d696923203ef.

Maybe it's worth trying out a copy of a large file to see when it starts to fail to verify everything.

jborean93 replied:
Ok I did some testing myself and can confirm those limits are still in place for SMB2. You can actually see the limit documented on the raw IOCTL docs for this command at https://docs.microsoft.com/en-us/windows/win32/devnotes/ioctl-copychunk.

The `Length` field states

> The number of bytes of data in the chunk to be copied. Must be greater than zero and less than or equal to 1 MB. Length * ChunkCount must be less than or equal to 16 MB.

So this confirms the max size of `1024 * 1024` for a chunk size and 16 chunks per request. Sorry for doubting you here :)

martinhoefling replied:
No worries. I have extracted it as constants and have added a proper description.

jborean93 commented on smbclient/_os.py:
```diff
@@ -1016,3 +1017,93 @@ def stat(self, follow_symlinks=True):
             return self._lstat
+
+
+def copyfile_server_side(src, dst, **kwargs):
```
All the other methods in `_os.py` are all server side but I've kept the names as just `rename`, `replace`. I think we should follow that here and just have this as `copyfile` and make sure we add in the docstring that it only works for copying on the same server/share.

martinhoefling replied:
done

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,178 @@
+    pass
+
+
+class SpecialFileError(EnvironmentError):
```
Doesn't seem to be used anymore

martinhoefling replied:
removed

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,178 @@
+    copyfile_server_side(src, dst, **kwargs)
+
+
+def copytree_copy(src, dst, symlinks=False, ignore=None, **kwargs):
```
Should this just be `copytree`?

martinhoefling replied:
Well, copytree under the hood uses copy2. I named it like that to make clear that it's not the same as shutil.copytree

jborean93 replied:
Hmm I see what you mean. I still think we should have it called copytree and just make it explicit in the docs that it just doesn’t copy the file metadata for now. We can always add that in a future release once we have a good and working copy2 implementation.

The reason why I think we should have this but not necessarily a copy2 for now is that it is better to have partial functionality whereas copy2’s only extra feature over just copy is something that I’m not 100% sure how to implement right now.

martinhoefling replied:
ok, renamed to copytree and added note in the docstring

jborean93 commented on smbclient/shutil.py:
```diff
@@ -0,0 +1,178 @@
+from smbprotocol.exceptions import SMBOSError
+
+
+class Error(EnvironmentError):
```
This seems quite generic, can we use one of the existing SMB Exceptions we have in `smbprotocol.exceptions`?

jborean93 commented on tests/test_smbclient_shutil.py:
```diff
@@ -0,0 +1,142 @@
+    with open_file("%s\\file1" % smb_share, mode='w') as fd:
+        fd.write(u"content")
+
+    sleep(0.1)
```
Why the sleep here?

jborean93 commented on tests/test_smbclient_shutil.py:
```diff
@@ -0,0 +1,142 @@
+    with open_file("%s\\file1" % smb_share, mode='w') as fd:
+        fd.write(u"content")
+
+    sleep(0.1)
```
Sleep again

martinhoefling replied:
removed

jborean93 commented on tests/test_smbclient_shutil.py:
```diff
@@ -0,0 +1,142 @@
+        fd.write(u"content")
+
+    sleep(0.1)
+    if (sys.version_info > (3, 0)):
```
Why the version difference?

martinhoefling replied:
That's a pytest bug. On python2, pytest cannot properly handle unicode error messages. I add a comment with a reference to the bug.

jborean93 replied:
That shouldn’t be an issue, the `smbprotocol._text` has a `to_native` function which can be used on Python 2 and 3 to create a native string from a Unicode string. Technically it’s a no-op on py3 but on py2 it should overcome the limitation of non-ASCII chars in an exception string.

You should just be able to do `match=re.escape(to_native(expected))` but best to test it out first.

martinhoefling replied:
With that change, I obtain the same error as before:
```
        expected = 'are the same file'
        with pytest.raises(ValueError, match=re.escape(to_native(expected))):
>           copy("%s\\file1" % smb_share, "%s\\file1" % smb_share)
E           UnicodeEncodeError: 'ascii' codec can't encode character u'\xfd' in position 30: ordinal not in range(128)
```

martinhoefling replied:
I was not able to resolve this for now. More ideas? I think the corresponding issue is: https://github.com/pytest-dev/pytest/issues/5478

jborean93 replied:
Make sure you do to_native when you raise the actual exception itself. Also it’s best to make sure your expectation is a Unicode string first by putting the u prefix before it.

If that doesn’t work we can leave it for now.

martinhoefling replied:
I tried
```
    if _samefile(src, dst, **kwargs):
        raise ValueError(to_native(u"`%s` and `%s` are the same file") % (src, dst))
```
Same error.

jborean93 replied:
No worries, I'll have a look at it later and will just keep it there if I can't fix it. Thanks for trying though.

jborean93 commented on tests/test_smbclient_shutil.py:
```diff
@@ -0,0 +1,142 @@
+    with open_file("%s\\file1" % smb_share, mode='w') as fd:
+        fd.write(u"content")
+
+    sleep(0.1)
```
Sleep

martinhoefling replied:
removed

jborean93 commented on tests/test_smbclient_shutil.py:
```diff
@@ -0,0 +1,142 @@
+    with open_file("%s\\dir2\\file2" % smb_share, mode='w') as fd:
+        fd.write(u"content")
+
+    sleep(0.01)
```
Sleep

jborean93 commented on tests/test_smbclient_shutil.py:
```diff
@@ -0,0 +1,142 @@
+    def ignore(src, names):
+        return [name for name in names if name == "file2"]
+
+    sleep(0.01)
```
Sleep

martinhoefling replied:
removed

jborean93 commented on smbclient/_os.py:
```diff
@@ -1016,3 +1017,93 @@ def stat(self, follow_symlinks=True):
+                      share_access='r',
+                      create_options=CreateOptions.FILE_NON_DIRECTORY_FILE, **kwargs) as raw_dst:
+
+            with SMBFileTransaction(raw_dst) as transaction_dst:
```
This line should be swapped with the for loop. Right now it will send a compound with multiple IOCTL requests which fails as soon as you go over 16 due to the `MAX_COPY_CHUNK_COUNT`. Swapping it means each request only contains up to 16MB of copy chunks hitting the max limit.

Also because of this you will have to move the transaction check below to the proper indentation.

martinhoefling replied:
swapped :-)

martinhoefling commented:
@jborean93 , I hope I have addressed most issues. Two are still open I think:
- SMBRawIO usage: Maybe you can take a look, why the suggestions didn't work out.
- pytest bug: Same thing. Maybe we just leave it like that and omit checking the message on python 2.7?

jborean93 commented:
I'll have a look at both points once easy-api has been merged into the master branch before finalising the new release as well.

martinhoefling commented:
@jborean93 thx, for all your comments and help. Lets get that one released and I would then work on copy2.

jborean93 commented:
@martinhoefling about those 2 remaining issues, here is the fix for the unicode chars in the exception https://github.com/jborean93/smbprotocol/commit/dd1b5de012144aacbaef6dcc7df14c7467c37333.

It wasn't an issue in pytest but rather how Python 2 handles native string compared to 3. In 2, a native string is actually a byte string and the default behaviour of converting a unicode string to byte is to encode it using the `ascii` encoding. This causes issues in cases where the unicode string contains non-ascii chars and then re-raises another ValueError for that. The way to get past that is to ensure you already pass in a native string (byte in 2, unicode in 3) so it doesn't try and encode it using ascii. The `to_native` method in smbprotocol handles all this for you and defaults to `utf-8` which is lot more sane in today's environment.

As for the SMBRawIO to open_file this was fixed in https://github.com/jborean93/smbprotocol/commit/39d82b7976292c19b9d8bfcbef7dd04125b97c43. Basically the SMBFileTransaction expects a SMBRawIO object and when you use open_file that can be encapsulated in a few different layers. In this case I am making sure I open the file in binary mode (`mode='*b'`) and I cna then access the raw IO object with the `raw` property. I can even take this a step futher and set `buffering=0` to `open_file` and that returns the relevant SMBRawIO object for the file type specified.

martinhoefling commented:
Thx!
